pkg/controller/elasticsearch/filesettings/secret.go

```diff
@@ -0,0 +1,191 @@
+	"encoding/json"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	commonv1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/common/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/elasticsearch/v1"
+	policyv1alpha1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/stackconfigpolicy/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/reconciler"
+	eslabel "github.com/elastic/cloud-on-k8s/v2/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/k8s"
+)
+
+const (
+	secureSettingsSecretsAnnotationName = "policy.k8s.elastic.co/secure-settings-secrets" //nolint:gosec
+	settingsHashAnnotationName          = "policy.k8s.elastic.co/settings-hash"
+	settingsSecretKey                   = "settings.json"
+)
+
+// NewSettingsSecret returns a new SettingsSecret for a given Elasticsearch and StackConfigPolicy.
+func NewSettingsSecret(version int64, current *SettingsSecret, es types.NamespacedName, policy *policyv1alpha1.StackConfigPolicy) (SettingsSecret, error) {
+	settings := NewEmptySettings(version)
+
+	// update the settings according to the config policy
+	if policy != nil {
+		err := settings.updateState(es, *policy)
+		if err != nil {
+			return SettingsSecret{}, err
+		}
+	}
+
+	// do not increment version if hash hasn't changed
+	if current != nil && !current.hasChanged(settings) {
+		version = current.Version
+		settings.Metadata.Version = current.Settings.Metadata.Version
+	}
+
+	// prepare the SettingsSecret
+	settingsBytes, err := json.Marshal(settings)
+	if err != nil {
+		return SettingsSecret{}, err
+	}
+	settingsSecret := SettingsSecret{
+		Secret: corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: es.Namespace,
+				Name:      esv1.FileSettingsSecretName(es.Name),
+				Labels:    eslabel.NewLabels(es),
+				Annotations: map[string]string{
+					settingsHashAnnotationName: settings.hash(),
+				},
+			},
+			Data: map[string][]byte{
+				settingsSecretKey: settingsBytes,
+			},
+		},
+		Settings: settings,
+		Version:  version,
+	}
+
+	if policy != nil {
+		// set this policy as soft owner of this Secret
+		settingsSecret.setSoftOwner(*policy)
+
+		// add the Secure Settings Secret sources to the Settings Secret
+		if err := settingsSecret.setSecureSettings(*policy); err != nil {
+			return SettingsSecret{}, err
+		}
+	}
+
+	return settingsSecret, nil
+}
+
+// NewSettingsSecretFromSecret returns a new Settings Secret from a given Secret.
+func NewSettingsSecretFromSecret(secret corev1.Secret) (SettingsSecret, error) {
+	var settings Settings
+	err := json.Unmarshal(secret.Data[settingsSecretKey], &settings)
+	if err != nil {
+		return SettingsSecret{}, err
+	}
+	version, err := strconv.ParseInt(settings.Metadata.Version, 10, 64)
+	if err != nil {
+		return SettingsSecret{}, err
+	}
+
+	return SettingsSecret{
+		Secret:   secret,
+		Settings: settings,
```

Comment:
If I see this correctly we only ever read this in tests, which is of course legitimate. But I wonder if we should do the parsing at all in production scope.

Reply:
Not exactly, at each reconciliation, we parse the content of the current Secret in the Settings struct to extract the current Settings versionand when the hash has not changed we compare this version to the actual version returned by the ES API. We can optimize a bit by doing the extraction only when the hash has not changed.
If we make `NewSettingsSecretWithVersion` return `(secret, version, error)`, the whole `SettingsSecret` struct becomes useless. It's probably simpler and clearer.